<?xml version="1.0" encoding="UTF-8"?><!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd"><html xmlns="http://www.w3.org/1999/xhtml" lang="en"><head><meta http-equiv="Content-Type" content="text/html;charset=UTF-8"/><link rel="stylesheet" href="../jacoco-resources/report.css" type="text/css"/><link rel="shortcut icon" href="../jacoco-resources/report.gif" type="image/gif"/><title>EffectsMetaData.java</title><link rel="stylesheet" href="../jacoco-resources/prettify.css" type="text/css"/><script type="text/javascript" src="../jacoco-resources/prettify.js"></script></head><body onload="window['PR_TAB_WIDTH']=4;prettyPrint()"><div class="breadcrumb" id="breadcrumb"><span class="info"><a href="../jacoco-sessions.html" class="el_session">Sessions</a></span><a href="../index.html" class="el_report">data-annotators</a> &gt; <a href="index.source.html" class="el_package">org.molgenis.data.annotation.core.effects</a> &gt; <span class="el_source">EffectsMetaData.java</span></div><h1>EffectsMetaData.java</h1><pre class="source lang-java linenums">package org.molgenis.data.annotation.core.effects;

import org.molgenis.data.annotation.web.meta.AnnotatorEntityType;
import org.molgenis.data.meta.model.Attribute;
import org.molgenis.data.meta.model.AttributeFactory;
import org.molgenis.data.vcf.utils.VcfWriterUtils;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Component;

import java.util.LinkedList;

import static org.molgenis.data.meta.AttributeType.STRING;
import static org.molgenis.data.meta.AttributeType.TEXT;

@Component
<span class="fc" id="L16">public class EffectsMetaData implements AnnotatorEntityType</span>
{
	@Autowired
	AttributeFactory attributeFactory;

	public static final String ID = &quot;id&quot;;
	public static final String VARIANT = VcfWriterUtils.VARIANT;
	public static final String ALT = &quot;Alt_Allele&quot;;
	public static final String ANNOTATION = &quot;Annotation&quot;;
	public static final String PUTATIVE_IMPACT = &quot;Putative_impact&quot;;
	public static final String GENE_NAME = &quot;Gene_Name&quot;;
	public static final String GENE_ID = &quot;Gene_ID&quot;;
	public static final String FEATURE_TYPE = &quot;Feature_type&quot;;
	public static final String FEATURE_ID = &quot;Feature_ID&quot;;
	public static final String TRANSCRIPT_BIOTYPE = &quot;Transcript_biotype&quot;;
	public static final String RANK_TOTAL = &quot;Rank_total&quot;;
	public static final String HGVS_C = &quot;HGVS_c&quot;;
	public static final String HGVS_P = &quot;HGVS_p&quot;;
	public static final String C_DNA_POSITION = &quot;cDNA_position&quot;;
	public static final String CDS_POSITION = &quot;CDS_position&quot;;
	public static final String PROTEIN_POSITION = &quot;Protein_position&quot;;
	public static final String DISTANCE_TO_FEATURE = &quot;Distance_to_feature&quot;;
	public static final String ERRORS = &quot;Errors&quot;;

	public LinkedList&lt;Attribute&gt; getOrderedAttributes()
	{
<span class="fc" id="L42">		Attribute ALT_ATTR = getAltAttr();</span>
<span class="fc" id="L43">		Attribute PUTATIVE_IMPACT_ATTR = getPutativeImpactAttr();</span>
<span class="fc" id="L44">		Attribute GENE_NAME_ATTR = getGeneNameAttr();</span>
<span class="fc" id="L45">		Attribute GENE_ID_ATTR = getGeneIdAttr();</span>
<span class="fc" id="L46">		Attribute FEATURE_TYPE_ATTR = getFeatureTypeAttr();</span>
<span class="fc" id="L47">		Attribute FEATURE_ID_ATTR = getFeatureIdAttr();</span>
<span class="fc" id="L48">		Attribute TRANSCRIPT_BIOTYPE_ATTR = getTranscriptBiotypeAttr();</span>
<span class="fc" id="L49">		Attribute RANK_TOTAL_ATTR = getRankTotalAttr();</span>
<span class="fc" id="L50">		Attribute HGVS_C_ATTR = getHgvsCAttr();</span>
<span class="fc" id="L51">		Attribute HGVS_P_ATTR = getHgvsPAttr();</span>
<span class="fc" id="L52">		Attribute C_DNA_POSITION_ATTR = getCdnaPositionAttr();</span>
<span class="fc" id="L53">		Attribute CDS_POSITION_ATTR = getCdsPositionAttr();</span>
<span class="fc" id="L54">		Attribute PROTEIN_POSITION_ATTR = getProteinPositionAttr();</span>
<span class="fc" id="L55">		Attribute DISTANCE_TO_FEATURE_ATTR = getDistanceToFeatureAttr();</span>
<span class="fc" id="L56">		Attribute ERRORS_ATTR = getErrorsAttr();</span>
<span class="fc" id="L57">		Attribute ANNOTATION_ATTR = getAnnotationAttr();</span>

<span class="fc" id="L59">		LinkedList&lt;Attribute&gt; attributes = new LinkedList&lt;&gt;();</span>
<span class="fc" id="L60">		attributes.add(ALT_ATTR);</span>
<span class="fc" id="L61">		attributes.add(GENE_NAME_ATTR);</span>
<span class="fc" id="L62">		attributes.add(ANNOTATION_ATTR);</span>
<span class="fc" id="L63">		attributes.add(PUTATIVE_IMPACT_ATTR);</span>
<span class="fc" id="L64">		attributes.add(GENE_ID_ATTR);</span>
<span class="fc" id="L65">		attributes.add(FEATURE_TYPE_ATTR);</span>
<span class="fc" id="L66">		attributes.add(FEATURE_ID_ATTR);</span>
<span class="fc" id="L67">		attributes.add(TRANSCRIPT_BIOTYPE_ATTR);</span>
<span class="fc" id="L68">		attributes.add(RANK_TOTAL_ATTR);</span>
<span class="fc" id="L69">		attributes.add(HGVS_C_ATTR);</span>
<span class="fc" id="L70">		attributes.add(HGVS_P_ATTR);</span>
<span class="fc" id="L71">		attributes.add(C_DNA_POSITION_ATTR);</span>
<span class="fc" id="L72">		attributes.add(CDS_POSITION_ATTR);</span>
<span class="fc" id="L73">		attributes.add(PROTEIN_POSITION_ATTR);</span>
<span class="fc" id="L74">		attributes.add(DISTANCE_TO_FEATURE_ATTR);</span>
<span class="fc" id="L75">		attributes.add(ERRORS_ATTR);</span>

<span class="fc" id="L77">		return attributes;</span>
	}

	public Attribute getTranscriptBiotypeAttr()
	{
<span class="fc" id="L82">		return attributeFactory.create()</span>
<span class="fc" id="L83">							   .setName(TRANSCRIPT_BIOTYPE)</span>
<span class="fc" id="L84">							   .setDataType(STRING)</span>
<span class="fc" id="L85">							   .setDescription(</span>
									   &quot;The bare minimum is at least a description on whether the transcript is {“Coding”, “Noncoding”}. Whenever possible, use ENSEMBL biotypes.(source:http://snpeff.sourceforge.net)&quot;)
<span class="fc" id="L87">							   .setAggregatable(true);</span>
	}

	public Attribute getRankTotalAttr()
	{
<span class="fc" id="L92">		return attributeFactory.create()</span>
<span class="fc" id="L93">							   .setName(RANK_TOTAL)</span>
<span class="fc" id="L94">							   .setDataType(STRING)</span>
<span class="fc" id="L95">							   .setDescription(</span>
									   &quot;Exon or Intron rank / total number of exons or introns(source:http://snpeff.sourceforge.net)&quot;);
	}

	public Attribute getHgvsCAttr()
	{
<span class="fc" id="L101">		return attributeFactory.create()</span>
<span class="fc" id="L102">							   .setName(HGVS_C)</span>
<span class="fc" id="L103">							   .setDataType(TEXT)</span>
<span class="fc" id="L104">							   .setDescription(</span>
									   &quot;Variant using HGVS notation (DNA level)(source:http://snpeff.sourceforge.net)&quot;);
	}

	public Attribute getHgvsPAttr()
	{
<span class="fc" id="L110">		return attributeFactory.create()</span>
<span class="fc" id="L111">							   .setName(HGVS_P)</span>
<span class="fc" id="L112">							   .setDataType(STRING)</span>
<span class="fc" id="L113">							   .setDescription(</span>
									   &quot;If variant is coding, this field describes the variant using HGVS notation (Protein level). Since transcript ID is already mentioned in ‘feature ID’, it may be omitted here.(source:http://snpeff.sourceforge.net)&quot;);
	}

	public Attribute getCdnaPositionAttr()
	{
<span class="fc" id="L119">		return attributeFactory.create()</span>
<span class="fc" id="L120">							   .setName(C_DNA_POSITION)</span>
<span class="fc" id="L121">							   .setDataType(STRING)</span>
<span class="fc" id="L122">							   .setDescription(</span>
									   &quot;Position in cDNA and trancript’s cDNA length (one based)(source:http://snpeff.sourceforge.net)&quot;);
	}

	public Attribute getCdsPositionAttr()
	{
<span class="fc" id="L128">		return attributeFactory.create()</span>
<span class="fc" id="L129">							   .setName(CDS_POSITION)</span>
<span class="fc" id="L130">							   .setDataType(STRING)</span>
<span class="fc" id="L131">							   .setDescription(</span>
									   &quot;Position and number of coding bases (one based includes START and STOP codons)(source:http://snpeff.sourceforge.net)&quot;);
	}

	public Attribute getProteinPositionAttr()
	{
<span class="fc" id="L137">		return attributeFactory.create()</span>
<span class="fc" id="L138">							   .setName(PROTEIN_POSITION)</span>
<span class="fc" id="L139">							   .setDataType(STRING)</span>
<span class="fc" id="L140">							   .setDescription(&quot;Position and number of AA (one based, including START, but not STOP)&quot;);</span>
	}

	public Attribute getDistanceToFeatureAttr()
	{
<span class="fc" id="L145">		return attributeFactory.create()</span>
<span class="fc" id="L146">							   .setName(DISTANCE_TO_FEATURE)</span>
<span class="fc" id="L147">							   .setDataType(STRING)</span>
<span class="fc" id="L148">							   .setDescription(</span>
									   &quot;All items in this field are options, so the field could be empty. Up/Downstream: Distance to first / last codon Intergenic: Distance to closest gene Distance to closest Intron boundary in exon (+/- up/downstream). If same, use positive number. Distance to closest exon boundary in Intron (+/- up/downstream) Distance to first base in MOTIF Distance to first base in miRNA Distance to exon-intron boundary in splice_site or splice _region ChipSeq peak: Distance to summit (or peak center) Histone mark / Histone state: Distance to summit (or peak center)(source:http://snpeff.sourceforge.net)&quot;);
	}

	public Attribute getErrorsAttr()
	{
<span class="fc" id="L154">		return attributeFactory.create()</span>
<span class="fc" id="L155">							   .setName(ERRORS)</span>
<span class="fc" id="L156">							   .setDataType(STRING)</span>
<span class="fc" id="L157">							   .setDescription(</span>
									   &quot;Add errors, warnings oErrors, Warnings or Information messages: Add errors, warnings or r informative message that can affect annotation accuracy. It can be added using either ‘codes’ (as shown in column 1, e.g. W1) or ‘message types’ (as shown in column 2, e.g. WARNING_REF_DOES_NOT_MATCH_GENOME). All these errors, warnings or information messages messages are optional.(source:http://snpeff.sourceforge.net)&quot;);
	}

	public Attribute getAnnotationAttr()
	{
<span class="fc" id="L163">		return attributeFactory.create()</span>
<span class="fc" id="L164">							   .setName(ANNOTATION)</span>
<span class="fc" id="L165">							   .setDataType(STRING)</span>
<span class="fc" id="L166">							   .setDescription(</span>
									   &quot;Annotated using Sequence Ontology terms. Multiple effects can be concatenated using ‘&amp;’ (source:http://snpeff.sourceforge.net)&quot;)
<span class="fc" id="L168">							   .setAggregatable(true);</span>
	}

	public Attribute getFeatureIdAttr()
	{
<span class="fc" id="L173">		return attributeFactory.create()</span>
<span class="fc" id="L174">							   .setName(FEATURE_ID)</span>
<span class="fc" id="L175">							   .setDataType(STRING)</span>
<span class="fc" id="L176">							   .setDescription(</span>
									   &quot;Depending on the annotation, this may be: Transcript ID (preferably using version number), Motif ID, miRNA, ChipSeq peak, Histone mark, etc. Note: Some features may not have ID (e.g. histone marks from custom Chip-Seq experiments may not have a unique ID).(source:http://snpeff.sourceforge.net)&quot;);
	}

	public Attribute getFeatureTypeAttr()
	{
<span class="fc" id="L182">		return attributeFactory.create()</span>
<span class="fc" id="L183">							   .setName(FEATURE_TYPE)</span>
<span class="fc" id="L184">							   .setDataType(STRING)</span>
<span class="fc" id="L185">							   .setDescription(</span>
									   &quot;Which type of feature is in the next field (e.g. transcript, motif, miRNA, etc.). It is preferred to use Sequence Ontology (SO) terms, but ‘custom’ (user defined) are allowed. ANN=A|stop_gained|HIGH|||transcript|... Tissue specific features may include cell type / tissue information separated by semicolon e.g.: ANN=A|histone_binding_site|LOW|||H3K4me3:HeLa-S3|...\n&quot;
											   + &quot;Feature ID: Depending on the annotation, this may be: Transcript ID (preferably using version number), Motif ID, miRNA, ChipSeq peak, Histone mark, etc. Note: Some features may not have ID (e.g. histone marks from custom Chip-Seq experiments may not have a unique ID). (source:http://snpeff.sourceforge.net)&quot;)
<span class="fc" id="L188">							   .setAggregatable(true);</span>
	}

	public Attribute getGeneIdAttr()
	{
<span class="fc" id="L193">		return attributeFactory.create()</span>
<span class="fc" id="L194">							   .setName(GENE_ID)</span>
<span class="fc" id="L195">							   .setDataType(STRING)</span>
<span class="fc" id="L196">							   .setDescription(&quot;Gene ID&quot;)</span>
<span class="fc" id="L197">							   .setAggregatable(true);</span>
	}

	public Attribute getGeneNameAttr()
	{
<span class="fc" id="L202">		return attributeFactory.create()</span>
<span class="fc" id="L203">							   .setName(GENE_NAME)</span>
<span class="fc" id="L204">							   .setDataType(STRING)</span>
<span class="fc" id="L205">							   .setDescription(</span>
									   &quot;Common gene name (HGNC). Optional: use closest gene when the variant is “intergenic”(source:http://snpeff.sourceforge.net)&quot;)
<span class="fc" id="L207">							   .setAggregatable(true);</span>
	}

	public Attribute getPutativeImpactAttr()
	{
<span class="fc" id="L212">		return attributeFactory.create()</span>
<span class="fc" id="L213">							   .setName(PUTATIVE_IMPACT)</span>
<span class="fc" id="L214">							   .setDataType(STRING)</span>
<span class="fc" id="L215">							   .setDescription(</span>
									   &quot;A simple estimation of putative impact / deleteriousness : {HIGH, MODERATE, LOW, MODIFIER}(source:http://snpeff.sourceforge.net)&quot;)
<span class="fc" id="L217">							   .setAggregatable(true);</span>
	}

	public Attribute getAltAttr()
	{
<span class="fc" id="L222">		return attributeFactory.create()</span>
<span class="fc" id="L223">							   .setName(ALT)</span>
<span class="fc" id="L224">							   .setDataType(TEXT)</span>
<span class="fc" id="L225">							   .setDescription(&quot;The alternative allele on which this EFFECT applies&quot;);</span>
	}
}
</pre><div class="footer"><span class="right">Created with <a href="http://www.jacoco.org/jacoco">JaCoCo</a> 0.8.1.201803210924</span></div></body></html>