<?xml version="1.0" encoding="UTF-8"?><!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd"><html xmlns="http://www.w3.org/1999/xhtml" lang="en"><head><meta http-equiv="Content-Type" content="text/html;charset=UTF-8"/><link rel="stylesheet" href="../jacoco-resources/report.css" type="text/css"/><link rel="shortcut icon" href="../jacoco-resources/report.gif" type="image/gif"/><title>FitConAnnotator.java</title><link rel="stylesheet" href="../jacoco-resources/prettify.css" type="text/css"/><script type="text/javascript" src="../jacoco-resources/prettify.js"></script></head><body onload="window['PR_TAB_WIDTH']=4;prettyPrint()"><div class="breadcrumb" id="breadcrumb"><span class="info"><a href="../jacoco-sessions.html" class="el_session">Sessions</a></span><a href="../index.html" class="el_report">data-annotators</a> &gt; <a href="index.source.html" class="el_package">org.molgenis.data.annotation.core.entity.impl</a> &gt; <span class="el_source">FitConAnnotator.java</span></div><h1>FitConAnnotator.java</h1><pre class="source lang-java linenums">package org.molgenis.data.annotation.core.entity.impl;

import org.molgenis.data.DataService;
import org.molgenis.data.Entity;
import org.molgenis.data.annotation.core.RepositoryAnnotator;
import org.molgenis.data.annotation.core.entity.AnnotatorConfig;
import org.molgenis.data.annotation.core.entity.AnnotatorInfo;
import org.molgenis.data.annotation.core.entity.EntityAnnotator;
import org.molgenis.data.annotation.core.entity.impl.framework.AbstractAnnotator;
import org.molgenis.data.annotation.core.entity.impl.framework.RepositoryAnnotatorImpl;
import org.molgenis.data.annotation.core.filter.MultiAllelicResultFilter;
import org.molgenis.data.annotation.core.query.LocusQueryCreator;
import org.molgenis.data.annotation.core.resources.Resource;
import org.molgenis.data.annotation.core.resources.Resources;
import org.molgenis.data.annotation.core.resources.impl.RepositoryFactory;
import org.molgenis.data.annotation.core.resources.impl.ResourceImpl;
import org.molgenis.data.annotation.core.resources.impl.SingleResourceConfig;
import org.molgenis.data.annotation.core.resources.impl.tabix.TabixRepositoryFactory;
import org.molgenis.data.annotation.web.settings.SingleFileLocationCmdLineAnnotatorSettingsConfigurer;
import org.molgenis.data.meta.model.Attribute;
import org.molgenis.data.meta.model.AttributeFactory;
import org.molgenis.data.meta.model.EntityType;
import org.molgenis.data.meta.model.EntityTypeFactory;
import org.molgenis.data.vcf.model.VcfAttributes;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;

import java.util.List;

import static com.google.common.collect.Lists.newArrayList;
import static org.molgenis.data.annotation.web.settings.FitConAnnotatorSettings.Meta.FITCON_LOCATION;
import static org.molgenis.data.meta.AttributeType.STRING;

@Configuration
<span class="nc" id="L36">public class FitConAnnotator implements AnnotatorConfig</span>
{
	public static final String NAME = &quot;fitcon&quot;;

	public static final String FITCON_SCORE = &quot;FITCON_SCORE&quot;;
	public static final String FITCON_SCORE_LABEL = &quot;FITCON_SCORE&quot;;
	public static final String FITCON_TABIX_RESOURCE = &quot;FitConTabixResource&quot;;

	@Autowired
	private Entity fitConAnnotatorSettings;

	@Autowired
	private DataService dataService;

	@Autowired
	private Resources resources;

	@Autowired
	private VcfAttributes vcfAttributes;

	@Autowired
	private EntityTypeFactory entityTypeFactory;

	@Autowired
	private AttributeFactory attributeFactory;

	private RepositoryAnnotatorImpl annotator;

	@Bean
	public RepositoryAnnotator fitcon()
	{
<span class="nc" id="L67">		annotator = new RepositoryAnnotatorImpl(NAME);</span>
<span class="nc" id="L68">		return annotator;</span>
	}

	@Override
	public void init()
	{
<span class="nc" id="L74">		List&lt;Attribute&gt; attributes = createFitconOutputAttributes();</span>

<span class="nc" id="L76">		AnnotatorInfo fitconInfo = AnnotatorInfo.create(AnnotatorInfo.Status.READY,</span>
				AnnotatorInfo.Type.EFFECT_PREDICTION, NAME,
				&quot;Summary: Annotating genetic variants, especially non-coding variants, &quot;
						+ &quot;for the purpose of identifying pathogenic variants remains a challenge. &quot;
						+ &quot;Combined annotation-dependent depletion (CADD) is an al- gorithm designed &quot;
						+ &quot;to annotate both coding and non-coding variants, and has been shown to &quot;
						+ &quot;outper- form other annotation algorithms. CADD trains a linear kernel support&quot;
						+ &quot; vector machine (SVM) to dif- ferentiate evolutionarily derived, likely benign,&quot;
						+ &quot; alleles from simulated, likely deleterious, variants. However, SVMs cannot &quot;
						+ &quot;capture non-linear relationships among the features, which can limit per- formance. &quot;
						+ &quot;To address this issue, we have developed FITCON. FITCON uses the same feature set and &quot;
						+ &quot;training data as CADD to train a deep neural network (DNN). DNNs can capture non-linear&quot;
						+ &quot; relation- ships among features and are better suited than SVMs for problems with a &quot;
						+ &quot;large number of samples and features. We exploit Compute Unified Device Architecture-compatible&quot;
						+ &quot; graphics processing units and deep learning techniques such as dropout and momentum training to&quot;
						+ &quot; accelerate the DNN train- ing. FITCON achieves about a 19%relative reduction in the error rate and&quot;
						+ &quot; about a 14%relative increase in the area under the curve (AUC) metric over CADDâ€™s SVMmethodology.&quot;
						+ &quot; All data and source code are available at https://cbcl.ics.uci.edu/ public_data/FITCON/. Contact:&quot;,
				attributes);
<span class="nc" id="L95">		EntityAnnotator entityAnnotator = new AbstractAnnotator(FITCON_TABIX_RESOURCE, fitconInfo,</span>
				new LocusQueryCreator(vcfAttributes), new MultiAllelicResultFilter(attributes, vcfAttributes),
				dataService, resources,
				new SingleFileLocationCmdLineAnnotatorSettingsConfigurer(FITCON_LOCATION, fitConAnnotatorSettings))
<span class="nc" id="L99">		{</span>
			@Override
			public List&lt;Attribute&gt; createAnnotatorAttributes(AttributeFactory attributeFactory)
			{
<span class="nc" id="L103">				return createFitconOutputAttributes();</span>
			}
		};

<span class="nc" id="L107">		annotator.init(entityAnnotator);</span>
<span class="nc" id="L108">	}</span>

	private List&lt;Attribute&gt; createFitconOutputAttributes()
	{
<span class="nc" id="L112">		List&lt;Attribute&gt; attributes = newArrayList();</span>
<span class="nc" id="L113">		Attribute fitcon_score = attributeFactory.create()</span>
<span class="nc" id="L114">												 .setName(FITCON_SCORE)</span>
<span class="nc" id="L115">												 .setDataType(STRING)</span>
<span class="nc" id="L116">												 .setDescription(</span>
														 &quot;fitness consequence score annotation of genetic variants using Fitcon scoring.&quot;)
<span class="nc" id="L118">												 .setLabel(FITCON_SCORE_LABEL);</span>

<span class="nc" id="L120">		attributes.add(fitcon_score);</span>
<span class="nc" id="L121">		return attributes;</span>
	}

	@Bean
	Resource fitconResource()
	{
		Resource fitConTabixResource;
<span class="nc" id="L128">		fitConTabixResource = new ResourceImpl(FITCON_TABIX_RESOURCE,</span>
				new SingleResourceConfig(FITCON_LOCATION, fitConAnnotatorSettings))
<span class="nc" id="L130">		{</span>

			@Override
			public RepositoryFactory getRepositoryFactory()
			{
<span class="nc" id="L135">				EntityType repoMetaData = entityTypeFactory.create(FITCON_TABIX_RESOURCE);</span>
<span class="nc" id="L136">				repoMetaData.addAttribute(vcfAttributes.getChromAttribute());</span>
<span class="nc" id="L137">				repoMetaData.addAttribute(vcfAttributes.getPosAttribute());</span>
<span class="nc" id="L138">				repoMetaData.addAttribute(vcfAttributes.getRefAttribute());</span>
<span class="nc" id="L139">				repoMetaData.addAttribute(attributeFactory.create().setName(&quot;Anc&quot;));</span>
<span class="nc" id="L140">				repoMetaData.addAttribute(vcfAttributes.getAltAttribute());</span>
<span class="nc" id="L141">				repoMetaData.addAttribute(attributeFactory.create().setName(&quot;Type&quot;));</span>
<span class="nc" id="L142">				repoMetaData.addAttribute(attributeFactory.create().setName(&quot;Length&quot;));</span>
<span class="nc" id="L143">				repoMetaData.addAttribute(attributeFactory.create().setName(&quot;isTv&quot;));</span>
<span class="nc" id="L144">				repoMetaData.addAttribute(attributeFactory.create().setName(&quot;isDerived&quot;));</span>
<span class="nc" id="L145">				repoMetaData.addAttribute(attributeFactory.create().setName(&quot;AnnoType&quot;));</span>
<span class="nc" id="L146">				repoMetaData.addAttribute(attributeFactory.create().setName(&quot;Consequence&quot;));</span>
<span class="nc" id="L147">				repoMetaData.addAttribute(attributeFactory.create().setName(&quot;ConsScore&quot;));</span>
<span class="nc" id="L148">				repoMetaData.addAttribute(attributeFactory.create().setName(&quot;ConsDetail&quot;));</span>
<span class="nc" id="L149">				repoMetaData.addAttribute(attributeFactory.create().setName(&quot;GC&quot;));</span>
<span class="nc" id="L150">				repoMetaData.addAttribute(attributeFactory.create().setName(&quot;CpG&quot;));</span>
<span class="nc" id="L151">				repoMetaData.addAttribute(attributeFactory.create().setName(&quot;mapAbility20bp&quot;));</span>
<span class="nc" id="L152">				repoMetaData.addAttribute(attributeFactory.create().setName(&quot;mapAbility35bp&quot;));</span>
<span class="nc" id="L153">				repoMetaData.addAttribute(attributeFactory.create().setName(&quot;scoreSegDup&quot;));</span>
<span class="nc" id="L154">				repoMetaData.addAttribute(attributeFactory.create().setName(&quot;priPhCons&quot;));</span>
<span class="nc" id="L155">				repoMetaData.addAttribute(attributeFactory.create().setName(&quot;mamPhCons&quot;));</span>
<span class="nc" id="L156">				repoMetaData.addAttribute(attributeFactory.create().setName(&quot;verPhCons&quot;));</span>
<span class="nc" id="L157">				repoMetaData.addAttribute(attributeFactory.create().setName(&quot;priPhyloP&quot;));</span>
<span class="nc" id="L158">				repoMetaData.addAttribute(attributeFactory.create().setName(&quot;mamPhyloP&quot;));</span>
<span class="nc" id="L159">				repoMetaData.addAttribute(attributeFactory.create().setName(&quot;verPhyloP&quot;));</span>
<span class="nc" id="L160">				repoMetaData.addAttribute(attributeFactory.create().setName(&quot;GerpN&quot;));</span>
<span class="nc" id="L161">				repoMetaData.addAttribute(attributeFactory.create().setName(&quot;GerpS&quot;));</span>
<span class="nc" id="L162">				repoMetaData.addAttribute(attributeFactory.create().setName(&quot;GerpRS&quot;));</span>
<span class="nc" id="L163">				repoMetaData.addAttribute(attributeFactory.create().setName(&quot;GerpRSpval&quot;));</span>
<span class="nc" id="L164">				repoMetaData.addAttribute(attributeFactory.create().setName(&quot;bStatistic&quot;));</span>
<span class="nc" id="L165">				repoMetaData.addAttribute(attributeFactory.create().setName(&quot;mutIndex&quot;));</span>
<span class="nc" id="L166">				repoMetaData.addAttribute(attributeFactory.create().setName(&quot;dnaHelT&quot;));</span>
<span class="nc" id="L167">				repoMetaData.addAttribute(attributeFactory.create().setName(&quot;dnaMGW&quot;));</span>
<span class="nc" id="L168">				repoMetaData.addAttribute(attributeFactory.create().setName(&quot;dnaProT&quot;));</span>
<span class="nc" id="L169">				repoMetaData.addAttribute(attributeFactory.create().setName(&quot;dnaRoll&quot;));</span>
<span class="nc" id="L170">				repoMetaData.addAttribute(attributeFactory.create().setName(&quot;mirSVR-Score&quot;));</span>
<span class="nc" id="L171">				repoMetaData.addAttribute(attributeFactory.create().setName(&quot;mirSVR-E&quot;));</span>
<span class="nc" id="L172">				repoMetaData.addAttribute(attributeFactory.create().setName(&quot;mirSVR-Aln&quot;));</span>
<span class="nc" id="L173">				repoMetaData.addAttribute(attributeFactory.create().setName(&quot;targetScan&quot;));</span>
				// fitcons can be NA so we need to catch the string value
<span class="nc" id="L175">				repoMetaData.addAttribute(attributeFactory.create().setName(&quot;FITCON_SCORE&quot;));</span>
<span class="nc" id="L176">				repoMetaData.addAttribute(attributeFactory.create().setName(&quot;cHmmTssA&quot;));</span>
<span class="nc" id="L177">				repoMetaData.addAttribute(attributeFactory.create().setName(&quot;cHmmTssAFlnk&quot;));</span>
<span class="nc" id="L178">				repoMetaData.addAttribute(attributeFactory.create().setName(&quot;cHmmTxFlnk&quot;));</span>
<span class="nc" id="L179">				repoMetaData.addAttribute(attributeFactory.create().setName(&quot;cHmmTx&quot;));</span>
<span class="nc" id="L180">				repoMetaData.addAttribute(attributeFactory.create().setName(&quot;cHmmTxWk&quot;));</span>
<span class="nc" id="L181">				repoMetaData.addAttribute(attributeFactory.create().setName(&quot;cHmmEnhG&quot;));</span>
<span class="nc" id="L182">				repoMetaData.addAttribute(attributeFactory.create().setName(&quot;cHmmEnh&quot;));</span>
<span class="nc" id="L183">				repoMetaData.addAttribute(attributeFactory.create().setName(&quot;cHmmZnfRpts&quot;));</span>
<span class="nc" id="L184">				repoMetaData.addAttribute(attributeFactory.create().setName(&quot;cHmmHet&quot;));</span>
<span class="nc" id="L185">				repoMetaData.addAttribute(attributeFactory.create().setName(&quot;cHmmTssBiv&quot;));</span>
<span class="nc" id="L186">				repoMetaData.addAttribute(attributeFactory.create().setName(&quot;cHmmBivFlnk&quot;));</span>
<span class="nc" id="L187">				repoMetaData.addAttribute(attributeFactory.create().setName(&quot;cHmmEnhBiv&quot;));</span>
<span class="nc" id="L188">				repoMetaData.addAttribute(attributeFactory.create().setName(&quot;cHmmReprPC&quot;));</span>
<span class="nc" id="L189">				repoMetaData.addAttribute(attributeFactory.create().setName(&quot;cHmmReprPCWk&quot;));</span>
<span class="nc" id="L190">				repoMetaData.addAttribute(attributeFactory.create().setName(&quot;cHmmQuies&quot;));</span>
<span class="nc" id="L191">				repoMetaData.addAttribute(attributeFactory.create().setName(&quot;EncExp&quot;));</span>
<span class="nc" id="L192">				repoMetaData.addAttribute(attributeFactory.create().setName(&quot;EncH3K27Ac&quot;));</span>
<span class="nc" id="L193">				repoMetaData.addAttribute(attributeFactory.create().setName(&quot;EncH3K4Me1&quot;));</span>
<span class="nc" id="L194">				repoMetaData.addAttribute(attributeFactory.create().setName(&quot;EncH3K4Me3&quot;));</span>
<span class="nc" id="L195">				repoMetaData.addAttribute(attributeFactory.create().setName(&quot;EncNucleo&quot;));</span>
<span class="nc" id="L196">				repoMetaData.addAttribute(attributeFactory.create().setName(&quot;EncOCC&quot;));</span>
<span class="nc" id="L197">				repoMetaData.addAttribute(attributeFactory.create().setName(&quot;EncOCCombPVal&quot;));</span>
<span class="nc" id="L198">				repoMetaData.addAttribute(attributeFactory.create().setName(&quot;EncOCDNasePVal&quot;));</span>
<span class="nc" id="L199">				repoMetaData.addAttribute(attributeFactory.create().setName(&quot;EncOCFairePVal&quot;));</span>
<span class="nc" id="L200">				repoMetaData.addAttribute(attributeFactory.create().setName(&quot;EncOCpolIIPVal&quot;));</span>
<span class="nc" id="L201">				repoMetaData.addAttribute(attributeFactory.create().setName(&quot;EncOCctcfPVal&quot;));</span>
<span class="nc" id="L202">				repoMetaData.addAttribute(attributeFactory.create().setName(&quot;EncOCmycPVal&quot;));</span>
<span class="nc" id="L203">				repoMetaData.addAttribute(attributeFactory.create().setName(&quot;EncOCDNaseSig&quot;));</span>
<span class="nc" id="L204">				repoMetaData.addAttribute(attributeFactory.create().setName(&quot;EncOCFaireSig&quot;));</span>
<span class="nc" id="L205">				repoMetaData.addAttribute(attributeFactory.create().setName(&quot;EncOCpolIISig&quot;));</span>
<span class="nc" id="L206">				repoMetaData.addAttribute(attributeFactory.create().setName(&quot;EncOCctcfSig&quot;));</span>
<span class="nc" id="L207">				repoMetaData.addAttribute(attributeFactory.create().setName(&quot;EncOCmycSig&quot;));</span>
<span class="nc" id="L208">				repoMetaData.addAttribute(attributeFactory.create().setName(&quot;Segway&quot;));</span>
<span class="nc" id="L209">				repoMetaData.addAttribute(attributeFactory.create().setName(&quot;tOverlapMotifs&quot;));</span>
<span class="nc" id="L210">				repoMetaData.addAttribute(attributeFactory.create().setName(&quot;motifDist&quot;));</span>
<span class="nc" id="L211">				repoMetaData.addAttribute(attributeFactory.create().setName(&quot;motifECount&quot;));</span>
<span class="nc" id="L212">				repoMetaData.addAttribute(attributeFactory.create().setName(&quot;motifEName&quot;));</span>
<span class="nc" id="L213">				repoMetaData.addAttribute(attributeFactory.create().setName(&quot;motifEHIPos&quot;));</span>
<span class="nc" id="L214">				repoMetaData.addAttribute(attributeFactory.create().setName(&quot;motifEScoreChng&quot;));</span>
<span class="nc" id="L215">				repoMetaData.addAttribute(attributeFactory.create().setName(&quot;TFBS&quot;));</span>
<span class="nc" id="L216">				repoMetaData.addAttribute(attributeFactory.create().setName(&quot;TFBSPeaks&quot;));</span>
<span class="nc" id="L217">				repoMetaData.addAttribute(attributeFactory.create().setName(&quot;TFBSPeaksMax&quot;));</span>
<span class="nc" id="L218">				repoMetaData.addAttribute(attributeFactory.create().setName(&quot;isKnownVariant&quot;));</span>
<span class="nc" id="L219">				repoMetaData.addAttribute(attributeFactory.create().setName(&quot;ESP_AF&quot;));</span>
<span class="nc" id="L220">				repoMetaData.addAttribute(attributeFactory.create().setName(&quot;ESP_AFR&quot;));</span>
<span class="nc" id="L221">				repoMetaData.addAttribute(attributeFactory.create().setName(&quot;ESP_EUR&quot;));</span>
<span class="nc" id="L222">				repoMetaData.addAttribute(attributeFactory.create().setName(&quot;TG_AF&quot;));</span>
<span class="nc" id="L223">				repoMetaData.addAttribute(attributeFactory.create().setName(&quot;TG_ASN&quot;));</span>
<span class="nc" id="L224">				repoMetaData.addAttribute(attributeFactory.create().setName(&quot;TG_AMR&quot;));</span>
<span class="nc" id="L225">				repoMetaData.addAttribute(attributeFactory.create().setName(&quot;TG_AFR&quot;));</span>
<span class="nc" id="L226">				repoMetaData.addAttribute(attributeFactory.create().setName(&quot;TG_EUR&quot;));</span>
<span class="nc" id="L227">				repoMetaData.addAttribute(attributeFactory.create().setName(&quot;minDistTSS&quot;));</span>
<span class="nc" id="L228">				repoMetaData.addAttribute(attributeFactory.create().setName(&quot;minDistTSE&quot;));</span>
<span class="nc" id="L229">				repoMetaData.addAttribute(attributeFactory.create().setName(&quot;GeneID&quot;));</span>
<span class="nc" id="L230">				repoMetaData.addAttribute(attributeFactory.create().setName(&quot;FeatureID&quot;));</span>
<span class="nc" id="L231">				repoMetaData.addAttribute(attributeFactory.create().setName(&quot;CCDS&quot;));</span>
<span class="nc" id="L232">				repoMetaData.addAttribute(attributeFactory.create().setName(&quot;GeneName&quot;));</span>
<span class="nc" id="L233">				repoMetaData.addAttribute(attributeFactory.create().setName(&quot;cDNApos&quot;));</span>
<span class="nc" id="L234">				repoMetaData.addAttribute(attributeFactory.create().setName(&quot;relcDNApos&quot;));</span>
<span class="nc" id="L235">				repoMetaData.addAttribute(attributeFactory.create().setName(&quot;CDSpos&quot;));</span>
<span class="nc" id="L236">				repoMetaData.addAttribute(attributeFactory.create().setName(&quot;relCDSpos&quot;));</span>
<span class="nc" id="L237">				repoMetaData.addAttribute(attributeFactory.create().setName(&quot;protPos&quot;));</span>
<span class="nc" id="L238">				repoMetaData.addAttribute(attributeFactory.create().setName(&quot;relProtPos&quot;));</span>
<span class="nc" id="L239">				repoMetaData.addAttribute(attributeFactory.create().setName(&quot;Domain&quot;));</span>
<span class="nc" id="L240">				repoMetaData.addAttribute(attributeFactory.create().setName(&quot;Dst2Splice&quot;));</span>
<span class="nc" id="L241">				repoMetaData.addAttribute(attributeFactory.create().setName(&quot;Dst2SplType&quot;));</span>
<span class="nc" id="L242">				repoMetaData.addAttribute(attributeFactory.create().setName(&quot;Exon&quot;));</span>
<span class="nc" id="L243">				repoMetaData.addAttribute(attributeFactory.create().setName(&quot;Intron&quot;));</span>
<span class="nc" id="L244">				repoMetaData.addAttribute(attributeFactory.create().setName(&quot;oAA&quot;));</span>
<span class="nc" id="L245">				repoMetaData.addAttribute(attributeFactory.create().setName(&quot;nAA&quot;));</span>
<span class="nc" id="L246">				repoMetaData.addAttribute(attributeFactory.create().setName(&quot;Grantham&quot;));</span>
<span class="nc" id="L247">				repoMetaData.addAttribute(attributeFactory.create().setName(&quot;PolyPhenCat&quot;));</span>
<span class="nc" id="L248">				repoMetaData.addAttribute(attributeFactory.create().setName(&quot;PolyPhenVal&quot;));</span>
<span class="nc" id="L249">				repoMetaData.addAttribute(attributeFactory.create().setName(&quot;SIFTcat&quot;));</span>
<span class="nc" id="L250">				repoMetaData.addAttribute(attributeFactory.create().setName(&quot;SIFTval&quot;));</span>
<span class="nc" id="L251">				repoMetaData.addAttribute(attributeFactory.create().setName(&quot;RawScore&quot;));</span>
<span class="nc" id="L252">				repoMetaData.addAttribute(attributeFactory.create().setName(&quot;PHRED&quot;));</span>

<span class="nc" id="L254">				Attribute idAttribute = attributeFactory.create().setName(&quot;id&quot;).setVisible(false).setIdAttribute(true);</span>
<span class="nc" id="L255">				repoMetaData.addAttribute(idAttribute);</span>

<span class="nc" id="L257">				return new TabixRepositoryFactory(repoMetaData);</span>
			}
		};

<span class="nc" id="L261">		return fitConTabixResource;</span>
	}
}
</pre><div class="footer"><span class="right">Created with <a href="http://www.jacoco.org/jacoco">JaCoCo</a> 0.8.1.201803210924</span></div></body></html>