<?xml version="1.0" encoding="UTF-8"?><!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd"><html xmlns="http://www.w3.org/1999/xhtml" lang="en"><head><meta http-equiv="Content-Type" content="text/html;charset=UTF-8"/><link rel="stylesheet" href="../jacoco-resources/report.css" type="text/css"/><link rel="shortcut icon" href="../jacoco-resources/report.gif" type="image/gif"/><title>SnpEffAnnotator.java</title><link rel="stylesheet" href="../jacoco-resources/prettify.css" type="text/css"/><script type="text/javascript" src="../jacoco-resources/prettify.js"></script></head><body onload="window['PR_TAB_WIDTH']=4;prettyPrint()"><div class="breadcrumb" id="breadcrumb"><span class="info"><a href="../jacoco-sessions.html" class="el_session">Sessions</a></span><a href="../index.html" class="el_report">data-annotators</a> &gt; <a href="index.source.html" class="el_package">org.molgenis.data.annotation.core.entity.impl.snpeff</a> &gt; <span class="el_source">SnpEffAnnotator.java</span></div><h1>SnpEffAnnotator.java</h1><pre class="source lang-java linenums">package org.molgenis.data.annotation.core.entity.impl.snpeff;

import org.molgenis.data.DataService;
import org.molgenis.data.Entity;
import org.molgenis.data.annotation.core.RepositoryAnnotator;
import org.molgenis.data.annotation.core.effects.EffectsMetaData;
import org.molgenis.data.annotation.core.entity.AnnotatorConfig;
import org.molgenis.data.vcf.model.VcfAttributes;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;

/**
 * SnpEff annotator
 * &lt;p&gt;
 * SnpEff is a genetic variant annotation and effect prediction toolbox. It annotates and predicts the effects of
 * variants on genes (such as amino acid changes). see http://snpeff.sourceforge.net/
 * &lt;p&gt;
 * For this annotator to work SnpEff.jar must be present on the filesystem at the location defined by the
 * RuntimeProperty 'snpeff_jar_location'
 * &lt;p&gt;
 * &lt;p&gt;
 * new ANN field replacing EFF:
 * &lt;p&gt;
 * ANN=A|missense_variant|MODERATE|NEXN|NEXN|transcript|NM_144573.3|Coding|8/13|c.733G&gt;A|p.Gly245Arg|1030/3389|733/2028|
 * 245/675||
 * &lt;p&gt;
 * &lt;p&gt;
 * -lof doesnt seem to work? would be great... http://snpeff.sourceforge.net/snpEff_lof_nmd.pdfs
 */
@Configuration
<span class="nc" id="L32">public class SnpEffAnnotator implements AnnotatorConfig</span>
{
	public static final String NAME = &quot;snpEff&quot;;

	@Autowired
	private SnpEffRunner snpEffRunner;

	@Autowired
	private Entity snpEffAnnotatorSettings;

	@Autowired
	private VcfAttributes vcfAttributes;

	@Autowired
	DataService dataService;

	@Autowired
	private EffectsMetaData effectsMetaData;
	private SnpEffRepositoryAnnotator annotator;

	@Bean
	public RepositoryAnnotator snpEff()
	{
<span class="nc" id="L55">		annotator = new SnpEffRepositoryAnnotator(NAME);</span>
<span class="nc" id="L56">		return annotator;</span>
	}

	@Override
	public void init()
	{
<span class="nc" id="L62">		annotator.init(snpEffRunner, snpEffAnnotatorSettings, vcfAttributes, effectsMetaData, dataService);</span>
<span class="nc" id="L63">	}</span>
}
</pre><div class="footer"><span class="right">Created with <a href="http://www.jacoco.org/jacoco">JaCoCo</a> 0.8.1.201803210924</span></div></body></html>