<?xml version="1.0" encoding="UTF-8"?><!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd"><html xmlns="http://www.w3.org/1999/xhtml" lang="en"><head><meta http-equiv="Content-Type" content="text/html;charset=UTF-8"/><link rel="stylesheet" href="../jacoco-resources/report.css" type="text/css"/><link rel="shortcut icon" href="../jacoco-resources/report.gif" type="image/gif"/><title>ClinvarMultiAllelicResultFilter.java</title><link rel="stylesheet" href="../jacoco-resources/prettify.css" type="text/css"/><script type="text/javascript" src="../jacoco-resources/prettify.js"></script></head><body onload="window['PR_TAB_WIDTH']=4;prettyPrint()"><div class="breadcrumb" id="breadcrumb"><span class="info"><a href="../jacoco-sessions.html" class="el_session">Sessions</a></span><a href="../index.html" class="el_report">data-annotators</a> &gt; <a href="index.source.html" class="el_package">org.molgenis.data.annotation.core.filter</a> &gt; <span class="el_source">ClinvarMultiAllelicResultFilter.java</span></div><h1>ClinvarMultiAllelicResultFilter.java</h1><pre class="source lang-java linenums">package org.molgenis.data.annotation.core.filter;

import com.google.common.base.Optional;
import com.google.common.collect.FluentIterable;
import org.molgenis.data.Entity;
import org.molgenis.data.MolgenisDataException;
import org.molgenis.data.annotation.core.entity.ResultFilter;
import org.molgenis.data.meta.model.Attribute;
import org.molgenis.data.vcf.model.VcfAttributes;

import java.util.*;

import static java.util.Objects.requireNonNull;

public class ClinvarMultiAllelicResultFilter implements ResultFilter
{
	private final VcfAttributes vcfAttributes;

	public ClinvarMultiAllelicResultFilter(VcfAttributes vcfAttributes)
<span class="fc" id="L20">	{</span>
<span class="fc" id="L21">		this.vcfAttributes = requireNonNull(vcfAttributes);</span>
<span class="fc" id="L22">	}</span>

	@Override
	public Collection&lt;Attribute&gt; getRequiredAttributes()
	{
<span class="nc" id="L27">		return Arrays.asList(vcfAttributes.getRefAttribute(), vcfAttributes.getAltAttribute());</span>
	}

	@Override
	public Optional&lt;Entity&gt; filterResults(Iterable&lt;Entity&gt; results, Entity annotatedEntity, boolean updateMode)
	{
<span class="pc bpc" id="L33" title="1 of 2 branches missed.">		if (updateMode == true)</span>
		{
<span class="nc" id="L35">			throw new MolgenisDataException(&quot;This annotator/filter does not support updating of values&quot;);</span>
		}
<span class="fc" id="L37">		Map&lt;String, String&gt; clnallValueMap = new LinkedHashMap&lt;&gt;();</span>
<span class="fc" id="L38">		Map&lt;String, String&gt; clnsigValueMap = new LinkedHashMap&lt;&gt;();</span>
<span class="fc" id="L39">		List&lt;Entity&gt; processedResults = new ArrayList&lt;&gt;();</span>

<span class="fc bfc" id="L41" title="All 2 branches covered.">		for (Entity entity : results)</span>
		{
<span class="pc bpc" id="L43" title="1 of 2 branches missed.">			if (entity.get(VcfAttributes.REF).equals(annotatedEntity.get(VcfAttributes.REF)))</span>
			{
<span class="fc" id="L45">				String[] alts = entity.getString(VcfAttributes.ALT).split(&quot;,&quot;);</span>
<span class="fc" id="L46">				String[] clnSigs = entity.getString(&quot;CLNSIG&quot;).split(&quot;,&quot;);</span>
<span class="fc" id="L47">				String[] clnAll = entity.getString(&quot;CLNALLE&quot;).split(&quot;,&quot;);</span>
<span class="fc" id="L48">				StringBuilder newClnlallAttributeValue = new StringBuilder();</span>
<span class="fc" id="L49">				StringBuilder newClnlsigAttributeValue = new StringBuilder();</span>
<span class="fc" id="L50">				String[] annotatedEntityAltAlleles = annotatedEntity.getString(VcfAttributes.ALT).split(&quot;,&quot;);</span>
				// sometimes the clnsig is not defined for all alternative alleles
				// so we need to check this and just add what we have
<span class="fc bfc" id="L53" title="All 2 branches covered.">				for (int i = 0; i &lt; clnSigs.length; i++)</span>
				{
<span class="fc" id="L55">					int significantAlleleIndex = Integer.parseInt(clnAll[i]);</span>

					// this means the no allele is associated with the gene of interest
<span class="pc bpc" id="L58" title="1 of 2 branches missed.">					if (significantAlleleIndex == -1) continue;</span>

						// this means the allele is based on the reference
<span class="pc bpc" id="L61" title="1 of 2 branches missed.">					else if (significantAlleleIndex == 0)</span>
					{

<span class="nc" id="L64">						String resultRefAllele = entity.getString(VcfAttributes.REF);</span>
<span class="nc" id="L65">						String refAllele = annotatedEntity.getString(VcfAttributes.REF);</span>

						// if annotated entity allele equals the clinvar significant allele we want it!
<span class="nc bnc" id="L68" title="All 2 branches missed.">						if (refAllele.equals(resultRefAllele))</span>
						{
							// if more than one clinsigs are available pair the right one with each allele
<span class="nc" id="L71">							clnallValueMap.put(refAllele, &quot;0&quot;);</span>
<span class="nc" id="L72">							clnsigValueMap.put(refAllele, clnSigs[i]);</span>

						}

<span class="nc" id="L76">					}</span>
					// 1 based so we need subtract 1 from the clnAll value
					else
					{

<span class="fc" id="L81">						significantAlleleIndex = significantAlleleIndex - 1;</span>

<span class="fc bfc" id="L83" title="All 2 branches covered.">						for (int j = 0; j &lt; annotatedEntityAltAlleles.length; j++)</span>
						{

<span class="fc bfc" id="L86" title="All 2 branches covered.">							if (annotatedEntityAltAlleles[j].equals(alts[significantAlleleIndex]))</span>
							{

<span class="fc" id="L89">								String newSignificantAlleleIndex = Integer.toString(j + 1);</span>

<span class="fc" id="L91">								clnallValueMap.put(alts[significantAlleleIndex], newSignificantAlleleIndex); // 1,2</span>
<span class="fc" id="L92">								clnsigValueMap.put(alts[significantAlleleIndex], clnSigs[i]); // &quot;5,4&quot;</span>

							}
						}
					}

				}

<span class="fc bfc" id="L100" title="All 2 branches covered.">				for (int i = 0; i &lt; annotatedEntityAltAlleles.length; i++)</span>
				{
<span class="fc bfc" id="L102" title="All 2 branches covered.">					if (i != 0)</span>
					{
<span class="fc" id="L104">						newClnlallAttributeValue.append(&quot;,&quot;);</span>
<span class="fc" id="L105">						newClnlsigAttributeValue.append(&quot;,&quot;);</span>
					}
<span class="fc bfc" id="L107" title="All 2 branches covered.">					if (clnallValueMap.get(annotatedEntityAltAlleles[i]) != null)</span>
					{
<span class="fc" id="L109">						newClnlallAttributeValue.append(clnallValueMap.get(annotatedEntityAltAlleles[i]));</span>
					}
					else
					{
						// missing allele in source, add a dot
<span class="fc" id="L114">						newClnlallAttributeValue.append(&quot;.&quot;);</span>
					}

<span class="fc bfc" id="L117" title="All 2 branches covered.">					if (clnsigValueMap.get(annotatedEntityAltAlleles[i]) != null)</span>
					{
<span class="fc" id="L119">						newClnlsigAttributeValue.append(clnsigValueMap.get(annotatedEntityAltAlleles[i]));</span>
					}
					else
					{
						// missing allele in source, add a dot
<span class="fc" id="L124">						newClnlsigAttributeValue.append(&quot;.&quot;);</span>
					}
				}
				// nothing found at all? result is empty
<span class="pc bpc" id="L128" title="1 of 2 branches missed.">				if (newClnlallAttributeValue.toString().equals(&quot;.&quot;))</span>
				{
<span class="nc" id="L130">					entity.set(&quot;CLNSIG&quot;, &quot;&quot;);</span>
<span class="nc" id="L131">					entity.set(&quot;CLNALLE&quot;, &quot;&quot;);</span>
				}
				else
				{

<span class="fc" id="L136">					entity.set(&quot;CLNALLE&quot;, newClnlallAttributeValue.toString());</span>
<span class="fc" id="L137">					entity.set(&quot;CLNSIG&quot;, newClnlsigAttributeValue.toString());</span>

				}

<span class="fc" id="L141">				processedResults.add(entity);</span>
			}

<span class="fc" id="L144">		}</span>

<span class="fc" id="L146">		return FluentIterable.from(processedResults).first();</span>
	}
}
</pre><div class="footer"><span class="right">Created with <a href="http://www.jacoco.org/jacoco">JaCoCo</a> 0.8.1.201803210924</span></div></body></html>